sshd/zenith/sshd/backends/crd.py

```diff
@@ -0,0 +1,182 @@
+                    return False
+        return True
+
+    def tunnel_init(
+        self,
+        subdomain: str,
+        host: str,
+        port: int,
+        ttl: int,
+        reap_after: int,
+        config_dict: t.Dict[str, t.Any]
+    ) -> str:
+        # Get the endpoints record, so that it can own our lease
+        # If it doesn't exist, create it
+        ekservices = self.ekclient.api(self.api_version).resource("services")
+        ekendpoints = self.ekclient.api(self.api_version).resource("endpoints")
+        try:
+            endpoints = ekendpoints.fetch(subdomain)
+        except ApiError as exc:
+            if exc.status_code == 404:
+                # Fetch the service so it can own the endpoints resource
+                service = ekservices.fetch(subdomain)
+                endpoints = ekendpoints.create(
+                    {
+                        "metadata": {
+                            "name": service["metadata"]["name"],
+                            "namespace": service["metadata"]["namespace"],
+                            "ownerReferences": [
+                                {
+                                    "apiVersion": service["apiVersion"],
+                                    "kind": service["kind"],
+                                    "name": service["metadata"]["name"],
+                                    "uid": service["metadata"]["uid"],
+                                    "blockOwnerDeletion": True,
+                                },
+                            ],
+                        }
+                    }
+                )
+            else:
+                raise
+
+        # Create a lease for the tunnel
+        ekleases = self.ekclient.api(self.api_version).resource("leases")
+        lease = ekleases.create(
+            {
+                "metadata": {
+                    # Generate a name based on the subdomain
+                    # We will use this as the tunnel ID
+                    "generateName": f"{subdomain}-",
+                    "ownerReferences": [
+                        {
+                            "apiVersion": endpoints["apiVersion"],
+                            "kind": endpoints["kind"],
+                            "name": endpoints["metadata"]["name"],
+                            "uid": endpoints["metadata"]["uid"],
+                            "blockOwnerDeletion": True,
+                        },
+                    ],
+                },
+                "spec": {
+                    "renewedAt": isotime(),
+                    "ttl": ttl,
+                    "reapAfter": reap_after,
+                }
+            }
+        )
+
+        # The tunnel ID is the lease name with the subdomain prefix removed
+        tunnel_id = lease["metadata"]["name"].removeprefix(f"{subdomain}-")
+
+        # Update the endpoints resource with the endpoint definition
+        # The endpoints resource should already exist
+        ekendpoints.patch(
```

Comment:
Can we add a warning here if the endpoints were not removed last time? I guess in the k8s case we expect multiples?

Reply:
Not entirely sure what you mean? Especially during a Kubernetes control plane upgrade, there could be more or less endpoints than we "expect" during "normal" operation. I'm not sure what meaningful check we could do here, given that the tunnels aren't supposed to have knowledge about their purpose or about each other.

Reply:
All we know at this point as that this is a new tunnel with a new ID that we are inserting.

Reply:
I really mean try to check for when the endpoint lists is growing really large, can that happen?